### PR TITLE
TRUNK-4939 VisitValidator should ensure that the visit start date can't be before the birth date of the patient 

### PR DESCRIPTION
**Pull request by tenas97**
## Description of what I changed
I added a test which validates birth date of patient to ensure it is not before the visit date of that same patient. I also included a 50% margin for error in cases where the birthdate is estimated.

## Issue I worked on
TRUNK-4939 VisitValidator should ensure that the visit start date can't be before the birth date of the patient 
https://issues.openmrs.org/browse/TRUNK-4939?filter=-1
## Checklist: I completed these to help reviewers :)

- [x] My pull request only contains **ONE single commit**
(the number above, next to the 'Commits' tab is 1).

  No? -> [read here](https://wiki.openmrs.org/display/docs/Pull+Request+Tips) on how to squash multiple commits into one

- [x] My IDE is configured to follow the [**code style**](https://wiki.openmrs.org/display/docs/Java+Conventions) of this project.

  No? Unsure? -> [configure your IDE](https://wiki.openmrs.org/display/docs/How-To+Setup+And+Use+Your+IDE), format the code and add the changes with `git add . && git commit --amend`

- [x] I have **added tests** to cover my changes. (If you refactored
  existing code that was well tested you do not have to add tests)

  No? -> write tests and add them to this commit `git add . && git commit --amend`

- [x] I ran `mvn clean package` right before creating this pull request and
  added all formatting changes to my commit.

  No? -> execute above command

- [x] All new and existing **tests passed**.

  No? -> figure out why and add the fix to your commit. It is your responsibility to make sure your code works.

- [x] My pull request is **based on the latest changes** of the master branch.

  No? Unsure? -> execute command `git pull --rebase upstream master`

### CODE REVIEW

**Comment by wluyima** [comment]
I see no unit test for you changes.
You might also want to look into why you have the auto format changes.

**Comment by tenas97** [comment]
@wluyima my change as a whole was a unit test to ensure that a patients birthdate isn't after a visit date attributed to that particular patient and also included a certain error margin to take care of patients who's birth dates are estimated

**Comment by tenas97** [comment]
About the auto changes, I think it came about after I ran mvn clean package.

File: api/src/test/java/org/openmrs/validator/VisitValidatorTest.java [hunk 1]
```diff
@@ -243,209 +243,241 @@ public void validate_shouldFailIfAnAttributeIsBad() {
...
+				.setVoidReason(
+						"too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text too long text");
+
 		Errors errors = new BindException(visit, "visit");
 		new VisitValidator().validate(visit, errors);
 		assertEquals(true, errors.hasFieldErrors("voidReason"));
 	}
+
+	@Test
+	public void validate_shouldFailIfStartDateTimeIsBeforeBirthDate() {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/validator/VisitValidatorTest.java]: You need to separate these tests i.e the one for estimated birthdate and the one for a non estimated date

File: api/src/test/java/org/openmrs/validator/VisitValidatorTest.java [hunk 2]
```diff
@@ -243,209 +243,241 @@ public void validate_shouldFailIfAnAttributeIsBad() {
...
+		Calendar calendar = Calendar.getInstance();
+		Visit visit = new Visit();
+		visit.setVisitType(visitService.getVisitType(1));
+		visit.setPatient(Context.getPatientService().getPatient(2));
+		visit.setStartDatetime(calendar.getTime());
+		Patient p = visit.getPatient();
+		if (p.getBirthdate().after(visit.getStartDatetime()) && !p.getBirthdateEstimated()) {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/validator/VisitValidatorTest.java]: You should not be having if clauses line this in a test, it's not good practice

File: api/src/test/java/org/openmrs/validator/VisitValidatorTest.java [hunk 3]
```diff
@@ -243,209 +243,241 @@ public void validate_shouldFailIfAnAttributeIsBad() {
...
+			Errors errors = new BindException(visit, "visit");
+			new VisitValidator().validate(visit, errors);
+			assertTrue (errors.hasFieldErrors("startDateTime"));
+		}
+
+		if (p.getBirthdateEstimated()){
+			if (p.getAge()> 2) {
```
Comment (wluyima) [inline api/src/test/java/org/openmrs/validator/VisitValidatorTest.java]: Can't have these if clauses, you need to write separate tests instead

Comment (tenas97) [inline api/src/test/java/org/openmrs/validator/VisitValidatorTest.java]: Is it possible for me to have just one if clause in the test to check if the birthdate is after a visit date for the said patient? Then I make another test which handles the estimated birthdate situation

Comment (wluyima) [inline api/src/test/java/org/openmrs/validator/VisitValidatorTest.java]: As in to make some sort of sanity check, right? If yes, that's what assertions are for in unit tests in place of if clauses

**Comment by tenas97** [comment]
@wluyima I am closing this pull request because I mistakenly deleted the source branch of this request and so I can't add commits to it. I recently opened another pull request for the same ticket with updated changes. Heres the link https://github.com/openmrs/openmrs-core/pull/2218

**Pull request closed without merging**